File: Two_pointers/Java/Valid_palindrome.java
// Time: O(n)
// Space: O(1)
// Problem link: https://leetcode.com/problems/valid-palindrome/description/

class Solution {
    public boolean isPalindrome(String s) {
        char[] str = s.toCharArray();
        int i = 0, j = s.length() - 1;

        while(i < j){
            while(i < j && !Character.isLetterOrDigit(str[i])){
                i++;
            }
            while(i < j && !Character.isLetterOrDigit(str[j])){
                j--;
            }
            if(Character.toLowerCase(str[i]) != Character.toLowerCase(str[j])){
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
}

File: Two_pointers/Python/Valid_palindrome.py
# Time: O(n)
# Space: O(1)
# Problem link: https://leetcode.com/problems/valid-palindrome/description/
class Solution:
    def isPalindrome(self, s: str) -> bool:
        result = ""
        for c in s:
            if c.isalnum():
                result += c.lower()
        return result == result[::-1]
